### PR TITLE
Ospi bus support for the stm32H7x devices

### PR DESCRIPTION
**Pull request by FRASTM**
This PR adds the support of the octoSPI bus for the stm32H7 devices
It updates the ospi flash controller driver
It also adds the OSPI instances to the stm32H7 mcus and boards

Not that DMA transfer is not yet in. There is no DMAMUX request for the octospi, only MDMA req 0x16, but the periph is not yet in.

requires the https://github.com/zephyrproject-rtos/zephyr/pull/45487

Signed-off-by: Francois Ramu <francois.ramu@st.com>

### CODE REVIEW

File: drivers/flash/flash_stm32_ospi.c [hunk 1]
```diff
@@ -1160,8 +1160,11 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	}
 
 	/* Initializes the independent peripherals clock */
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
-
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_PLL);
+#else
+	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Comment (manoj153) [inline drivers/flash/flash_stm32_ospi.c:1167]: should we refactor to 
```
#if defined(CONFIG_SOC_SERIES_STM32H7X)
#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_PLL
#else
#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_SYSCLK
#endif
```

and then

```
__HAL_RCC_OSPI_CONFIG(OSPI_CLK_SRC);
```

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32_ospi.c [hunk 2]
```diff
@@ -935,8 +935,8 @@ static void flash_stm32_ospi_isr(const struct device *dev)
 	HAL_OSPI_IRQHandler(&dev_data->hospi);
 }
 
-/* weak function required for HAL compilation */
-__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
+/* weak function required for HAL compilation */__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:939]:
```suggestion
#if !defined(CONFIG_SOC_SERIES_STM32H7X)
/* weak function required for HAL compilation */
__weak HAL_StatusTypeDef HAL_DMA_Abort_IT(DMA_HandleTypeDef *hdma)
```
Is slightly easier to read ;)

**Comment by FRASTM** [comment]
rebase and fix merge conflicts

File: drivers/flash/flash_stm32_ospi.c (hunk 2 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: yes, changed

File: drivers/flash/flash_stm32_ospi.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:1168]: Ok, It measns that the OSPI clock will comes from the  board DTS, overwriting the property as follows:
```
 &octospi1 {
	/delete-property/ clocks;
	clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00004000>,
		<&rcc STM32_SRC_PLL1_Q OSPI_SEL(1)>;
```
And then the flash_stm32_ospi_init function of the ospi driver will configure the clock with the clock_control_on function.
--> I propose another PR for that change

**Comment by FRASTM** [comment]
rebase
add the OSPI_CLK_SRC to configure the independent peripheral clock (until the  https://github.com/zephyrproject-rtos/zephyr/pull/46703 is completed)
Note the compilation warning  on #define FLASH_SECTOR_SIZE        4096 --> fixed by https://github.com/zephyrproject-rtos/zephyr/pull/45487

File: drivers/flash/flash_stm32_ospi.c (hunk 1 above)
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: done

**Comment by FRASTM** [comment]
tested on the stm32h735g_disco with  west build -p auto -b stm32h735g_disco samples/drivers/spi_flash
```

*** Booting Zephyr OS build zephyr-v3.1.0-589-ge9bd6e5b4bee  ***
JEDEC OSPI-NOR SPI flash testing                                            
==========================  
Test 1: Flash erase       
Flash erase succeeded! 

Test 2: Flash write
Attempting to write 4 bytes
Data read matches data written. Good!! 
```

File: drivers/flash/flash_stm32_ospi.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: DNM: waiting for the alternate clock model

**Comment by FRASTM** [comment]
rebase after merge of https://github.com/zephyrproject-rtos/zephyr/pull/46842 
Align on the https://github.com/zephyrproject-rtos/zephyr/pull/46703  to define the alternate clock source in the DTS of the soc for each octospi peripheral to be **pll1_q**
The default clock source (0) for the octospi is not the PCLK1 (but _rcc_hclk3_) so it must be defined explicitly.

**Comment by carlescufi** [comment]
@manoj153 please revisit

**Review by GeorgeCGV — DISMISSED** [review]
LGTM

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32_ospi.c [hunk 3]
```diff
@@ -1569,7 +1576,7 @@ static int flash_stm32_ospi_init(const struct device *dev)
 	}
 
 	/* Initializes the independent peripherals clock */
-	__HAL_RCC_OSPI_CONFIG(RCC_OSPICLKSOURCE_SYSCLK); /* */
+	__HAL_RCC_OSPI_CONFIG(OSPI_CLK_SRC);
```
[the remark this answers is not in the document]
Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c]: rebased

File: drivers/flash/flash_stm32_ospi.c [hunk 4]
```diff
@@ -41,6 +41,12 @@ LOG_MODULE_REGISTER(flash_stm32_ospi, CONFIG_FLASH_LOG_LEVEL);
 /* used as default value for DTS writeoc */
 #define SPI_NOR_WRITEOC_NONE 0xFF
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_PLL
+#else
+#define OSPI_CLK_SRC RCC_OSPICLKSOURCE_SYSCLK
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
+
```
Comment (GeorgeCGV) [inline drivers/flash/flash_stm32_ospi.c:49]:
```suggestion
```

Can be removed as it is not used anymore.

Comment (FRASTM) [reply drivers/flash/flash_stm32_ospi.c:49]: line removed

**Comment by FRASTM** [comment]
rebase on 60271524087bb4382c8b4db0f2e17a87e4120065
remove unused OSPI_CLK_SRC definition

**Review by GeorgeCGV — APPROVED** [review]
LGTM

**Review by erwango — APPROVED** [review]

File: boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts [hunk 5]
```diff
@@ -224,3 +224,44 @@
...
+		     &octospim_p1_io4_pc1 &octospim_p1_io5_ph3
+		     &octospim_p1_io6_pg9 &octospim_p1_io7_pd7
+		     &octospim_p1_dqs_pc5>;
+	pinctrl-names = "default";
+
+	status = "okay";
+
+	mx25lm51245: ospi-nor-flash@0 {
+		compatible = "st,stm32-ospi-nor";
+		label = "MX25LM51245";
```
Comment (fabiobaltieri) [inline boards/arm/stm32h7b3i_dk/stm32h7b3i_dk.dts:241]: Label property are being wiped out all over from the repository, https://github.com/zephyrproject-rtos/zephyr/pull/47881 wiped them from this binding. Can you drop them from here as well?

**Pull request merged**